File: Day06/p2.py
with open("customs_answers.txt", "r") as file:
    answers = [answer.split('\n') for answer in file.read().split("\n\n")]

for i, answer in enumerate(answers):
    union = set(answer[0]).intersection(*answer[1:])
    answers[i] = union

total = 0
for answer in answers:
    total += len(answer)

print(total)

File: Day06/p2.hs
module P2 where

import Data.List
import P1 (Group, parseFile)

main :: IO ()
main = do
    groups <- parseFile "customs_answers.txt"
    let totalAnswers = countAnswers groups
    putStrLn $ "Total answers: " ++ show totalAnswers


countAnswers :: (Num a) => [Group] -> a
countAnswers = sum . map countGroupAnswers

countGroupAnswers :: (Num a) => Group -> a
countGroupAnswers groupAnswers = foldl' isUnanimous 0 allAnswers
    where allAnswers = group . sort . concat $ groupAnswers
          numActors = length groupAnswers
          isUnanimous acc g = 
              if length g == numActors then acc + 1 else acc

